kingpin/actors/aws/elb.py

```diff
@@ -506,11 +525,65 @@ def _remove(self, elb, instances):
             elb: boto Loadbalancer object
             instances: list of instance ids.
         """
-        yield self.thread(elb.deregister_instances, instances)
+        self.log.info(('Removing instances from %s: %s'
+                      % (elb, ', '.join(instances))))
+
+        if not self._dry:
+            yield self.thread(elb.deregister_instances, instances)
+            yield self._wait_on_draining(elb)
+            self.log.info('Done.')
+
+    @gen.coroutine
+    def _wait_on_draining(self, elb):
+        """Waits for the ELB Connection Draining to occur.
+
+        ELB Connection Draining is a configured-setting on the ELB that will
+        continue to allow existing connections to be handeled before finally
+        cutting them off at the timeout. This method will detect if connection
+        draining is enabled, and optionally "sleep" for that time period before
+        returning from the actor.
+
+        Args:
+            elb: boto Loadbalancer object
+        """
+        if not self.option('wait_on_draining'):
+            self.log.warning('Not waiting for connections to drain!')
+
+        attrs = yield self.thread(elb.get_attributes)
+        if attrs.connection_draining.enabled:
+            timeout = attrs.connection_draining.timeout
+
+            self.log.info('Connection Draining Enabled, waiting %s(s)'
+                          % timeout)
+            yield utils.tornado_sleep(timeout)
+
+    @gen.coroutine
+    def _find_instance_elbs(self, instances):
+        """Finds all ELBs that Instances are members of.
+
+        Searches through all of the ELBs in a particular region and looks for
+        which ones have any of the instances supplied in them. Creates a list
+        of the ELBs, and returns the entire list.
+
+        Args:
+            instances: A list of Instance IDs
+
+        Returns:
+            a list of LoadBalancer objects
+        """
+        all_elbs = yield self.thread(self.elb_conn.get_all_load_balancers)
+        elbs_with_members = []
+
+        for instance in instances:
+            elbs = filter(lambda lb: instance in [i.id for i in lb.instances],
+                          all_elbs)
+            self.log.debug('%s is a member of %s' % (instance, elbs))
+            elbs_with_members.extend(elbs)
```

Comment:
Kinda funny that we're not leveraging the generator functionality of python here. Instead of fetching them all and returning a list we could actually yield one elb at a time.

Not worth changing things IMO because we've been doing this for a while now